supervisely/api/image_api.py

```diff
@@ -1606,38 +1673,81 @@ def _upload_bulk_add(
+            zip(
+                batched(names, batch_size=batch_size),
+                batched(items, batch_size=batch_size),
+                batched(metas, batch_size=batch_size),
             )
-            if progress_cb is not None:
-                progress_cb(len(images))
+        ):
+            for retry in range(2):
+                images = get_images(batch_names, batch_items, batch_metas)
+                try:
+                    response = self._api.post(
+                        "images.bulk.add",
+                        {
+                            ApiField.DATASET_ID: dataset_id,
+                            ApiField.IMAGES: images,
+                            ApiField.FORCE_METADATA_FOR_LINKS: force_metadata_for_links,
+                            ApiField.SKIP_VALIDATION: skip_validation,
+                        },
+                    )
+                    if progress_cb is not None:
+                        progress_cb(len(images))
+
+                    for info_json in response.json():
+                        info_json_copy = info_json.copy()
+                        if info_json.get(ApiField.MIME, None) is not None:
+                            info_json_copy[ApiField.EXT] = info_json[ApiField.MIME].split("/")[1]
+                        # results.append(self.InfoType(*[info_json_copy[field_name] for field_name in self.info_sequence()]))
+                        results.append(self._convert_json_info(info_json_copy))
+                    break
+                except HTTPError as e:
+                    error_details = e.response.json().get("details")
+                    if (
+                        conflict_resolution is not None
+                        and e.response.status_code == 400
+                        and error_details.get("type") == "NONUNIQUE"
+                    ):
+                        logger.info(
+                            f"Handling the exception above with '{conflict_resolution}' conflict resolution method"
+                        )
 
-            for info_json in response.json():
-                info_json_copy = info_json.copy()
-                if info_json.get(ApiField.MIME, None) is not None:
-                    info_json_copy[ApiField.EXT] = info_json[ApiField.MIME].split("/")[1]
-                # results.append(self.InfoType(*[info_json_copy[field_name] for field_name in self.info_sequence()]))
-                results.append(self._convert_json_info(info_json_copy))
+                        error_names = [error["name"] for error in error_details["errors"]]
+                        ids = [error["id"] for error in error_details["errors"]]
+
+                        if conflict_resolution == "replace":
+                            logger.info(f"Image ids to be removed: {ids}")
+                            self._api.image.remove_batch(ids)
+                            continue
+
+                        batch_names_copy = batch_names.copy()
```

Comment:
Let's start optimizations from something like this:

```python
                except HTTPError as e:
                    error_details = e.response.json().get("details")
                    if (
                        conflict_resolution is not None
                        and e.response.status_code == 400
                        and error_details.get("type") == "NONUNIQUE"
                    ):
                        logger.info(
                            f"Handling the exception above with '{conflict_resolution}' conflict resolution method"
                        )
                        errors: List[Dict] = error_details.get("errors", [])

                        if conflict_resolution == "replace":
                            ids_to_remove = [error["id"] for error in errors]
                            logger.info(f"Follower ids will be removed: {ids_to_remove}")
                            self._api.image.remove_batch(ids_to_remove)
                            continue
                        
                        name_to_index = {name: idx for idx, name in enumerate(batch_names)}
                        errors = sorted(errors, key=lambda x: name_to_index[x["name"]], reverse=True)
    
                        if conflict_resolution == "rename":
                            for error in errors:
                                error_image_name = error["name"]
                                batch_names[batch_names[name_to_index[error_image_name]]] = generate_name(error_image_name)                    
                                
                        elif conflict_resolution == "skip":
                            for error in errors:
                                error_image_name = error["name"]
                                error_index = name_to_index[error_image_name]

                                idx_to_id[error_index + batch_count * batch_size] = error["id"]
                                for l in [batch_items, batch_metas, batch_names]:
                                    l.pop(error_index)
```
                                    
We need to get rid of the `.index()` method and use dict mapping.